### PR TITLE
Creating reports automatically

### PR DESCRIPTION
**Pull request by WilsonReese**
Created an action to generate a report for all of the bench users automatically. I will use Heroku's Scheduler tool so a period will be automatically be created every week, and thus all the reports will be created.

### CODE REVIEW

File: app/controllers/periods_controller.rb [hunk 1]
```diff
@@ -66,4 +67,14 @@ def set_period
...
     end
+
+    # Create reports for all bench users for a given period
+    def create_reports_for_bench_users(period)
+      User.bench_users.each do |user|
+        Report.create!(
+          user: user,
+          period: period
+        )
+      end
```
Comment (jelaniwoods) [inline app/controllers/periods_controller.rb]: This feels something that needs to happen anytime a Period is created, does that sound right?

If so, then I think this fits better as [an active record callback](https://guides.rubyonrails.org/active_record_callbacks.html) instead of a method in a controller. Controllers are usually only responsible for methods that handle requests directly (using `render` or `redirect_to`) or methods that assist with that process by defining instance variables that are necessary for View templates.

But let's say someone created a `Period` in the `rails console`— `create_reports_for_bench_users` wouldn't run because `PeriodsController#create` never ran.

Using callbacks, you could configure Periods such that after any Period is created ActiveRecord will run a method for you. In this case you can make the method `create_reports_for_bench_users` and not remove this logic from the controller

**Review by jelaniwoods — COMMENTED** [review]
@WilsonReese nice work! I appreciate the PR description and the smaller size of this PR 🙌🏽 

I left a comment about Callback methods, that I think will be useful to read and know about. Let me know if you have any questions.

File: app/controllers/periods_controller.rb (hunk 1 above)
Comment (WilsonReese) [reply app/controllers/periods_controller.rb]: Hey Jelani, that is great feedback. Yes, this is supposed to happen any time a period is created. Active record callbacks are incredible - so helpful. I am kicking myself that I did not understand it until now. Makes sense to me to move it to the models file instead of the controller.

I have made an update on the same branch and pushed it. I then tried to and make a new pull request, but I didn't have that option. I am still trying to get used to pull requests and code review. But I think the changes updated in this pull request without having to make a new one. Let me know if how I did this works and is the best way to do it moving forward.

Comment (jelaniwoods) [reply app/controllers/periods_controller.rb]: @WilsonReese You can think of the pull request flow as something like:
1. You push a branch you plan to merge to GitHub and open a PR
2. You ask for review on said PR
3. Someone reviews the PR
4. You make changes based off of feedback and push them to the open PR
5. You ask for review again (steps 3-5 can sometimes repeat)
6. The reviewer approves all the changes you've made and you merge into `main` branch.

So in short, you did the right thing! Does that help?

File: app/controllers/periods_controller.rb [hunk 2]
```diff
@@ -25,6 +25,7 @@ def create
 
     respond_to do |format|
       if @period.save
+        # create_reports_for_bench_users(@period)
```
Comment (jelaniwoods) [inline app/controllers/periods_controller.rb]: This is safe to remove now that you have a callback in the Period model. If you wanted to keep the comment to refer to this version of the code, you still have the Git commit you can always jump back to

https://github.com/WilsonReese-AppDev/status-reporting/blob/f2b201d1a57436e4f2a3770c825b27bee58a9bfb/app/controllers/periods_controller.rb#L23-L36

File: app/models/period.rb [hunk 3]
```diff
@@ -12,4 +12,17 @@
...
+
+  private
+      # Create reports for all bench users for a given period
+    def create_reports_for_bench_users
+      User.bench_users.each do |user|
+        Report.create!(
+          user: user,
+          period: self
+        )
+      end
```
Comment (jelaniwoods) [inline app/models/period.rb]: Since you have an association between `User` and `Report`, you can condense this a little to be something like

```rb
      User.bench_users.each do |user|
        user.reports.create!(period: self)
      end
```
Since you also have an associate between Period and Reports you could do
```rb
      User.bench_users.each do |user|
        self.reports.create!(user: user)
      end
```

I could buy arguments for either one. You might prefer the `user.reports` way because you're looping over _users_ or you might prefer the `self.reports` way because you're in the `Period` class. 

---

Unrelated to this PR, but I noticed you defined the scope `bench_users`

https://github.com/WilsonReese-AppDev/status-reporting/blob/f2b201d1a57436e4f2a3770c825b27bee58a9bfb/app/models/user.rb#L30

With Rails 6, you [automatically get some scopes defined for Enum columns as well as some other things](https://edgeapi.rubyonrails.org/classes/ActiveRecord/Enum.html). So you should automatically have a scope available to you called `on_bench` that is identical to the `bench_users` scope.

**Review by jelaniwoods — COMMENTED** [review]
@WilsonReese 👏🏽 I left a few more comments for you to look at!

File: app/controllers/periods_controller.rb (hunk 1 above)
Comment (WilsonReese) [reply app/controllers/periods_controller.rb]: Yes that helps. It is good to know that pushing to an open PR works!

File: app/models/period.rb (hunk 3 above)
Comment (WilsonReese) [reply app/models/period.rb]: Amazing. Scopes were still a bit fuzzy to me, so I was looking for an opportunity to make them work. But it is great to know that enums automatically define scopes. I plan to add a couple more enums to my models, so I will make use of that functionality!
